Version-less_number_of_pull_get_power/doc/man/man5/slurm.conf.5
.TH "slurm.conf" "5" "February 2015" "slurm.conf 15.08" "Slurm configuration file"

.SH "NAME"
slurm.conf \- Slurm configuration file

.SH "DESCRIPTION"
\fBslurm.conf\fP is an ASCII file which describes general SLURM
configuration information, the nodes to be managed, information about
how those nodes are grouped into partitions, and various scheduling
parameters associated with those partitions. This file should be
consistent across all nodes in the cluster.
.LP
The file location can be modified at system build time using the
DEFAULT_SLURM_CONF parameter or at execution time by setting the SLURM_CONF
environment variable. The SLURM daemons also allow you to override
both the built\-in and environment\-provided location using the "\-f"
option on the command line.
.LP
The contents of the file are case insensitive except for the names of nodes
and partitions. Any text following a "#" in the configuration file is treated
as a comment through the end of that line.
Changes to the configuration file take effect upon restart of
SLURM daemons, daemon receipt of the SIGHUP signal, or execution
of the command "scontrol reconfigure" unless otherwise noted.
.LP
If a line begins with the word "Include" followed by whitespace
and then a file name, that file will be included inline with the current
configuration file. For large or complex systems, multiple configuration files
may prove easier to manage and enable reuse of some files (See INCLUDE
MODIFIERS for more details).
.LP
Note on file permissions:
.LP
The \fIslurm.conf\fR file must be readable by all users of SLURM, since it
is used by many of the SLURM commands.  Other files that are defined
in the \fIslurm.conf\fR file, such as log files and job accounting files,
may need to be created/owned by the user "SlurmUser" to be successfully
accessed.  Use the "chown" and "chmod" commands to set the ownership
and permissions appropriately.
See the section \fBFILE AND DIRECTORY PERMISSIONS\fR for information
about the various files and directories used by SLURM.

.SH "PARAMETERS"
.LP
The overall configuration parameters available include:

.TP
\fBAccountingStorageBackupHost\fR
The name of the backup machine hosting the accounting storage database.
If used with the accounting_storage/slurmdbd plugin, this is where the backup
slurmdbd would be running.
Only used for database type storage plugins, ignored otherwise.

.TP
\fBAccountingStorageEnforce\fR
This controls what level of association\-based enforcement to impose
on job submissions.  Valid options are any combination of
\fIassociations\fR, \fIlimits\fR, \fInojobs\fR, \fInosteps\fR, \fIqos\fR, \fIsafe\fR, and \fIwckeys\fR, or
\fIall\fR for all things (expect nojobs and nosteps, they must be requested as well).

If limits, qos, or wckeys are set, associations will automatically be set.

If wckeys is set, TrackWCKey will automatically be set.

If safe is set, limits and associations will automatically be set.

If nojobs is set nosteps will automatically be set.

By enforcing Associations no new job is allowed to run unless a corresponding
association exists in the system.  If limits are enforced users can be
limited by association to whatever job size or run time limits are defined.

If nojobs is set Slurm will not account for any jobs or steps on the system,
like wise if nosteps is set Slurm will not account for any steps ran limits
will still be enforced.

If safe is enforced a job will only be launched against an association or qos
that has a GrpCPUMins limit set if the job will be able to run to completion.
Without this option set, jobs will be launched as long as their usage
hasn't reached the cpu-minutes limit which can lead to jobs being
launched but then killed when the limit is reached.

With qos and/or wckeys enforced jobs will not be scheduled unless a valid qos
and/or workload characterization key is specified.

When \fBAccountingStorageEnforce\fR is changed, a restart of the slurmctld
daemon is required (not just a "scontrol reconfig").

.TP
\fBAccountingStorageHost\fR
The name of the machine hosting the accounting storage database.
Only used for database type storage plugins, ignored otherwise.
Also see \fBDefaultStorageHost\fR.

.TP
\fBAccountingStorageLoc\fR
The fully qualified file name where accounting records are written
when the \fBAccountingStorageType\fR is "accounting_storage/filetxt"
or else the name of the database where accounting records are stored when the
\fBAccountingStorageType\fR is a database.
Also see \fBDefaultStorageLoc\fR.

.TP
\fBAccountingStoragePass\fR
The password used to gain access to the database to store the
accounting data.  Only used for database type storage plugins, ignored
otherwise.  In the case of SLURM DBD (Database Daemon) with MUNGE
authentication this can be configured to use a MUNGE daemon
specifically configured to provide authentication between clusters
while the default MUNGE daemon provides authentication within a
cluster.  In that case, \fBAccountingStoragePass\fR should specify the
named port to be used for communications with the alternate MUNGE
daemon (e.g.  "/var/run/munge/global.socket.2"). The default value is
NULL.  Also see \fBDefaultStoragePass\fR.

.TP
\fBAccountingStoragePort\fR
The listening port of the accounting storage database server.
Only used for database type storage plugins, ignored otherwise.
Also see \fBDefaultStoragePort\fR.

.TP
\fBAccountingStorageType\fR
The accounting storage mechanism type.  Acceptable values at
present include "accounting_storage/filetxt",
"accounting_storage/mysql", "accounting_storage/none"
and "accounting_storage/slurmdbd".  The
"accounting_storage/filetxt" value indicates that accounting records
will be written to the file specified by the
\fBAccountingStorageLoc\fR parameter.  The "accounting_storage/mysql"
value indicates that accounting records will be written to a MySQL or
MariaDB database specified by the \fBAccountingStorageLoc\fR parameter.
The "accounting_storage/slurmdbd" value indicates that accounting records
will be written to the SLURM DBD, which manages an underlying MySQL
database. See "man slurmdbd" for more information.  The
default value is "accounting_storage/none" and indicates that account
records are not maintained.
Note: The filetxt plugin records only a limited subset of accounting
information and will prevent some sacct options from proper operation.
Also see \fBDefaultStorageType\fR.

.TP
\fBAccountingStorageUser\fR
The user account for accessing the accounting storage database.
Only used for database type storage plugins, ignored otherwise.
Also see \fBDefaultStorageUser\fR.

.TP
\fBAccountingStoreJobComment\fR
If set to "YES" then include the job's comment field in the job
complete message sent to the Accounting Storage database.  The default
is "YES".

.TP
\fBAcctGatherNodeFreq\fR
The AcctGather plugins sampling interval for node accounting.
For AcctGather plugin values of none, this parameter is ignored.
For all other values this parameter is the number
of seconds between node accounting samples. For the
acct_gather_energy/rapl plugin, set a value less
than 300 because the counters may overflow beyond this rate.
The default value is zero. This value disables accounting sampling
for nodes. Note: The accounting sampling interval for jobs is
determined by the value of \fBJobAcctGatherFrequency\fR.

.TP
\fBAcctGatherEnergyType\fR
Identifies the plugin to be used for energy consumption accounting.
The jobacct_gather plugin and slurmd daemon call this plugin to collect
energy consumption data for jobs and nodes. The collection of energy
consumption data takes place on node level, hence only in case of exclusive
job allocation the energy consumption measurements will reflect the jobs
real consumption. In case of node sharing between jobs the reported consumed
energy per job (through sstat or sacct) will not reflect the real energy
consumed by the jobs.

Configurable values at present are:
.RS
.TP 20
\fBacct_gather_energy/none\fR
No energy consumption data is collected.
.TP
\fBacct_gather_energy/ipmi\fR
Energy consumption data is collected from the Baseboard Management Controller
(BMC) using the Intelligent Platform Management Interface (IPMI).
.TP
\fBacct_gather_energy/rapl\fR
Energy consumption data is collected from hardware sensors using the Running
Average Power Limit (RAPL) mechanism. Note that enabling RAPL may require the
execution of the command "sudo modprobe msr".
.RE

.TP
\fBAcctGatherInfinibandType\fR
Identifies the plugin to be used for infiniband network traffic accounting.
The plugin is activated only when profiling on hdf5 files is activated and
the user asks for network data collection for jobs through \-\-profile=Network
(or =All). The collection of network traffic data takes place on node level,
hence only in case of exclusive job allocation the collected values will
reflect the jobs real traffic. All network traffic data are logged on hdf5 files
per job on each node. No storage on the Slurm database takes place.

Configurable values at present are:
.RS
.TP 20
\fBacct_gather_infiniband/none\fR
No infiniband network data are collected.
.TP
\fBacct_gather_infiniband/ofed\fR
Infiniband network traffic data are collected from the hardware monitoring
counters of Infiniband devices through the OFED library.
.RE

.TP
\fBAcctGatherFilesystemType\fR
Identifies the plugin to be used for filesystem traffic accounting.
The plugin is activated only when profiling on hdf5 files is activated and
the user asks for filesystem data collection for jobs through \-\-profile=Lustre
(or =All). The collection of filesystem traffic data takes place on node level,
hence only in case of exclusive job allocation the collected values will
reflect the jobs real traffic. All filesystem traffic data are logged on hdf5 files
per job on each node. No storage on the Slurm database takes place.

Configurable values at present are:
.RS
.TP 20
\fBacct_gather_filesystem/none\fR
No filesystem data are collected.
.TP
\fBacct_gather_filesystem/lustre\fR
Lustre filesystem traffic data are collected from the counters found in
/proc/fs/lustre/.
.RE

.TP
\fBAcctGatherProfileType\fR
Identifies the plugin to be used for detailed job profiling.
The jobacct_gather plugin and slurmd daemon call this plugin to collect
detailed data such as I/O counts, memory usage, or energy consumption for jobs
and nodes. There are interfaces in this plugin to collect data as step start
and completion, task start and completion, and at the account gather
frequency. The data collected at the node level is related to jobs only in
case of exclusive job allocation.

Configurable values at present are:
.RS
.TP 20
\fBacct_gather_profile/none\fR
No profile data is collected.
.TP
\fBacct_gather_profile/hdf5\fR
This enables the HDF5 plugin. The directory where the profile files
are stored and which values are collected are configured in the
acct_gather.conf file.
.RE

.TP
\fBAllowSpecResourcesUsage\fR
If set to 1, Slurm allows individual jobs to override node's configured
CoreSpecCount value. For a job to take advantage of this feature,
a command line option of \-\-core\-spec must be specified.  The default
value for this option is 1 for Cray systems and 0 for other system types.

.TP
\fBAuthInfo\fR
Additional information to be used for authentication of communications
between the Slurm daemons (slurmctld and slurmd) and the Slurm
clients.  The interpretation of this option is specific to the
configured \fBAuthType\fR.
Multiple options may be specified in a comma delimited list.
If not specified, the default authentication information will be used.
.RS
.TP 14
\fBcred_expire\fR
Default job step credential lifetime, in seconds (e.g. "cred_expire=1200").
It must be sufficiently long enough to load user environment, run prolog,
deal with the slurmd getting paged out of memory, etc.
This also controls how long a requeued job must wait before starting again.
The default value is 120 seconds.
.TP
\fBsocket\fR
Path name to a MUNGE daemon socket to use
(e.g. "socket=/var/run/munge/munge.socket.2").
The default value is "/var/run/munge/munge.socket.2".
Used by \fIauth/munge\fR and \fIcrypto/munge\fR.
.TP
\fBttl\fR
Credential lifetime, in seconds (e.g. "ttl=300").
The default value is dependent upon the Munge installation, but is typically
300 seconds.
.RE

.TP
\fBAuthType\fR
The authentication method for communications between SLURM
components.
Acceptable values at present include "auth/none", "auth/authd",
and "auth/munge".
The default value is "auth/munge".
"auth/none" includes the UID in each communication, but it is not verified.
This may be fine for testing purposes, but
\fBdo not use "auth/none" if you desire any security\fR.
"auth/authd" indicates that Brett Chun's authd is to be used (see
"http://www.theether.org/authd/" for more information. Note that
authd is no longer actively supported).
"auth/munge" indicates that LLNL's MUNGE is to be used
(this is the best supported authentication mechanism for SLURM,
see "http://munge.googlecode.com/" for more information).
All SLURM daemons and commands must be terminated prior to changing
the value of \fBAuthType\fR and later restarted (SLURM jobs can be
preserved).

.TP
\fBBackupAddr\fR
The name that \fBBackupController\fR should be referred to in
establishing a communications path. This name will
be used as an argument to the gethostbyname() function for
identification. For example, "elx0000" might be used to designate
the Ethernet address for node "lx0000".
By default the \fBBackupAddr\fR will be identical in value to
\fBBackupController\fR.

.TP
\fBBackupController\fR
The name of the machine where SLURM control functions are to be
executed in the event that \fBControlMachine\fR fails. This node
may also be used as a compute server if so desired. It will come into service
as a controller only upon the failure of ControlMachine and will revert
to a "standby" mode when the ControlMachine becomes available once again.
This should be a node name without the full domain name.   I.e., the hostname
returned by the \fIgethostname()\fR function cut at the first dot (e.g. use
"tux001" rather than "tux001.my.com").
The backup controller recovers state information from the
\fBStateSaveLocation\fR directory, which must be readable and writable from both
the primary and backup controllers.
While not essential, it is recommended that you specify a backup controller.
See  the \fBRELOCATING CONTROLLERS\fR section if you change this.

.TP
\fBBatchStartTimeout\fR
The maximum time (in seconds) that a batch job is permitted for
launching before being considered missing and releasing the
allocation. The default value is 10 (seconds). Larger values may be
required if more time is required to execute the \fBProlog\fR, load
user environment variables (for Moab spawned jobs), or if the slurmd
daemon gets paged from memory.
.br
.br
\fBNote\fR: The test for a job being successfully launched is only performed when
the Slurm daemon on the compute node registers state with the slurmctld daemon
on the head node, which happens fairly rarely.
Therefore a job will not necessarily be terminated if its start time exceeds
\fBBatchStartTimeout\fR.
This configuration parameter is also applied to launch tasks and avoid aborting
\fBsrun\fR commands due to long running \fBProlog\fR scripts.

.TP
\fBBurstBufferType\fR
The plugin used to manage burst buffers.
Acceptable values at present include "burst_buffer/none".
More information later...

.TP
\fBCacheGroups\fR
If set to 1, the slurmd daemon will cache /etc/groups entries.
This can improve performance for highly parallel jobs if NIS servers
are used and unable to respond very quickly.
The default value is 0 to disable caching group data.

.TP
\fBCheckpointType\fR
The system\-initiated checkpoint method to be used for user jobs.
The slurmctld daemon must be restarted for a change in \fBCheckpointType\fR
to take effect. Supported values presently include:
.RS
.TP 18
\fBcheckpoint/aix\fR
for IBM AIX systems only
.TP
\fBcheckpoint/blcr\fR
Berkeley Lab Checkpoint Restart (BLCR).
NOTE: If a file is found at sbin/scch (relative to the SLURM installation
location), it will be executed upon completion of the checkpoint. This can
be a script used for managing the checkpoint files.
NOTE: SLURM's BLCR logic only supports batch jobs.
.TP
\fBcheckpoint/none\fR
no checkpoint support (default)
.TP
\fBcheckpoint/ompi\fR
OpenMPI (version 1.3 or higher)
.TP
\fBcheckpoint/poe\fR
for use with IBM POE (Parallel Operating Environment) only
.RE

.TP
\fBChosLoc\fR
If configured, then any processes invoked on the user behalf (namely the
SPANK prolog/epilog scripts and the slurmstepd processes, which in turn spawn
the user batch script and applications) are not directly executed by the slurmd
daemon, but instead the \fBChosLoc\fR program is executed.
That program's argument are the program and arguments that would otherwise be
invoked directly by the slurmd daemon.
The intent of this feature is to be able to run a user application in some
sort of container.
This option specified the fully qualified pathname of the chos command
(see https://github.com/scanon/chos for details).

.TP
\fBClusterName\fR
The name by which this SLURM managed cluster is known in the
accounting database.  This is needed distinguish accounting records
when multiple clusters report to the same database. Because of limitations
in some databases, any upper case letters in the name will be silently mapped
to lower case. In order to avoid confusion, it is recommended that the name
be lower case.


.TP
\fBCompleteWait\fR
The time, in seconds, given for a job to remain in COMPLETING state
before any additional jobs are scheduled.
If set to zero, pending jobs will be started as soon as possible.
Since a COMPLETING job's resources are released for use by other
jobs as soon as the \fBEpilog\fR completes on each individual node,
this can result in very fragmented resource allocations.
To provide jobs with the minimum response time, a value of zero is
recommended (no waiting).
To minimize fragmentation of resources, a value equal to \fBKillWait\fR
plus two is recommended.
In that case, setting \fBKillWait\fR to a small value may be beneficial.
The default value of \fBCompleteWait\fR is zero seconds.
The value may not exceed 65533.

.TP
\fBControlAddr\fR
Name that \fBControlMachine\fR should be referred to in
establishing a communications path. This name will
be used as an argument to the gethostbyname() function for
identification. For example, "elx0000" might be used to designate
the Ethernet address for node "lx0000".
By default the \fBControlAddr\fR will be identical in value to
\fBControlMachine\fR.

.TP
\fBControlMachine\fR
The short hostname of the machine where SLURM control functions are
executed (i.e. the name returned by the command "hostname \-s", use
"tux001" rather than "tux001.my.com").
This value must be specified.
In order to support some high availability architectures, multiple
hostnames may be listed with comma separators and one \fBControlAddr\fR
must be specified. The high availability system must insure that the
slurmctld daemon is running on only one of these hosts at a time.
See the \fBRELOCATING CONTROLLERS\fR section if you change this.

.TP
\fBCoreSpecPlugin\fR
Identifies the plugins to be used for enforcement of core specialization.
The slurmd daemon must be restarted for a change in CoreSpecPlugin
to take effect.
Acceptable values at present include:
.RS
.TP 20
\fBcore_spec/cray\fR
used only for Cray systems
.TP
\fBcore_spec/none\fR
used for all other system types
.RE

.TP
\fBCpuFreqDef\fR
Default CPU frequency to be set when no jobs are running.
The CPU frequency can also be set to this value after a catastrophic failure
when state information has been lost.
Acceptable values at present include:
.RS
.TP 14
\fBLow\fR
the lowest available frequency
.TP
\fBHigh\fR
the highest available frequency
.TP
\fBHighM1\fR
(high minus one) will select the next highest available frequency
.TP
\fBMedium\fR
attempts to set a frequency in the middle of the available range
.TP
\fBConservative\fR
attempts to use the Conservative CPU governor
.TP
\fBOnDemand\fR
attempts to use the OnDemand CPU governor (the default value)
.TP
\fBPerformance\fR
attempts to use the Performance CPU governor
.TP
\fBPowerSave\fR
attempts to use the PowerSave CPU governor
.RE

.TP
\fBCpuFreqGovernors\fR
List of CPU frequency governors allowed to be set with the salloc, sbatch, or
srun option  \-\-cpu\-freq.
Acceptable values at present include:
.RS
.TP 14
\fBConservative\fR
attempts to use the Conservative CPU governor
.TP
\fBOnDemand\fR
attempts to use the OnDemand CPU governor (the default value)
.TP
\fBPerformance\fR
attempts to use the Performance CPU governor
.TP
\fBPowerSave\fR
attempts to use the PowerSave CPU governor
.TP
\fBUserSpace\fR
attempts to use the UserSpace CPU governor
.RE
The default is OnDemand.
.TP
\fBCryptoType\fR
The cryptographic signature tool to be used in the creation of
job step credentials.
The slurmctld daemon must be restarted for a change in \fBCryptoType\fR
to take effect.
Acceptable values at present include "crypto/munge" and "crypto/openssl".
The default value is "crypto/munge".

.TP
\fBDebugFlags\fR
Defines specific subsystems which should provide more detailed event logging.
Multiple subsystems can be specified with comma separators.
Most DebugFlags will result in verbose logging for the identified subsystems
and could impact performance.  The below DB_* flags are only useful when
writing directly to the database.  If using the DBD put these debug flags in the
slurmdbd.conf.
Valid subsystems available today (with more to come) include:
.RS
.TP 17
\fBBackfill\fR
Backfill scheduler details
.TP
\fBBackfillMap\fR
Backfill scheduler to log a very verbose map of reserved resources through
time. Combine with \fBBackfill\fR for a verbose and complete view of the
backfill scheduler's work.
.TP
\fBBGBlockAlgo\fR
BlueGene block selection details
.TP
\fBBGBlockAlgoDeep\fR
BlueGene block selection, more details
.TP
\fBBGBlockPick\fR
BlueGene block selection for jobs
.TP
\fBBGBlockWires\fR
BlueGene block wiring (switch state details)
.TP
\fBBurstBuffer\fR
Burst Buffer plugin
.TP
\fBCPU_Bind\fR
CPU binding details for jobs and steps
.TP
\fBCpuFrequency\fR
Cpu frequency details for jobs and steps using the \-\-cpu\-freq option.
.TP
\fBDB_ASSOC\fR
SQL statements/queries when dealing with associations in the database.
.TP
\fBDB_EVENT\fR
SQL statements/queries when dealing with (node) events in the database.
.TP
\fBDB_JOB\fR
SQL statements/queries when dealing with jobs in the database.
.TP
\fBDB_QOS\fR
SQL statements/queries when dealing with QOS in the database.
.TP
\fBDB_QUERY\fR
SQL statements/queries when dealing with transactions and such in the database.
.TP
\fBDB_RESERVATION\fR
SQL statements/queries when dealing with reservations in the database.
.TP
\fBDB_RESOURCE\fR
SQL statements/queries when dealing with resources like licenses in the
database.
.TP
\fBDB_STEP\fR
SQL statements/queries when dealing with steps in the database.
.TP
\fBDB_USAGE\fR
SQL statements/queries when dealing with usage queries and inserts
in the database.
.TP
\fBDB_WCKEY\fR
SQL statements/queries when dealing with wckeys in the database.
\fBEnergy\fR
AcctGatherEnergy debug info
.TP
\fBExtSensors\fR
External Sensors debug info
.TP
\fBFrontEnd\fR
Front end node details
.TP
\fBGres\fR
Generic resource details
.TP
\fBGang\fR
Gang scheduling details
.TP
\fBJobContainer\fR
Job container plugin details
.TP
\fBLicense\fB
License management details
.TP
\fBNO_CONF_HASH\fR
Do not log when the slurm.conf files differs between SLURM daemons
.TP
\fBPower\fR
Power management plugin
.TP
\fBPriority\fB
Job prioritization
.TP
\fBProtocol\fR
Communication protocol details
.TP
\fBReservation\fB
Advanced reservations
.TP
\fBSelectType\fR
Resource selection plugin
.TP
\fBSICP\fR
Inter\-cluster job details
.TP
\fBSteps\fR
Slurmctld resource allocation for job steps
.TP
\fBSwitch\fR
Switch plugin
.TP
\fBTraceJobs\fR
Trace jobs in slurmctld. It will print detailed job information
including state, job ids and allocated nodes counter.
.TP
\fBTriggers\fR
Slurmctld triggers
.TP
\fBWiki\fR
Sched/wiki and wiki2 communications
.RE

.TP
\fBDefMemPerCPU\fR
Default real memory size available per allocated CPU in MegaBytes.
Used to avoid over\-subscribing memory and causing paging.
\fBDefMemPerCPU\fR would generally be used if individual processors
are allocated to jobs (\fBSelectType=select/cons_res\fR).
The default value is 0 (unlimited).
Also see \fBDefMemPerNode\fR and \fBMaxMemPerCPU\fR.
\fBDefMemPerCPU\fR and \fBDefMemPerNode\fR are mutually exclusive.
NOTE: Enforcement of memory limits currently requires enabling of
accounting, which samples memory use on a periodic basis (data need
not be stored, just collected).

.TP
\fBDefMemPerNode\fR
Default real memory size available per allocated node in MegaBytes.
Used to avoid over\-subscribing memory and causing paging.
\fBDefMemPerNode\fR would generally be used if whole nodes
are allocated to jobs (\fBSelectType=select/linear\fR) and
resources are shared (\fBShared=yes\fR or \fBShared=force\fR).
The default value is 0 (unlimited).
Also see \fBDefMemPerCPU\fR and \fBMaxMemPerNode\fR.
\fBDefMemPerCPU\fR and \fBDefMemPerNode\fR are mutually exclusive.
NOTE: Enforcement of memory limits currently requires enabling of
accounting, which samples memory use on a periodic basis (data need
not be stored, just collected).

.TP
\fBDefaultStorageHost\fR
The default name of the machine hosting the accounting storage and
job completion databases.
Only used for database type storage plugins and when the
\fBAccountingStorageHost\fR and \fBJobCompHost\fR have not been
defined.

.TP
\fBDefaultStorageLoc\fR
The fully qualified file name where accounting records and/or job
completion records are written when the \fBDefaultStorageType\fR is
"filetxt" or the name of the database where accounting records and/or job
completion records are stored when the \fBDefaultStorageType\fR is a
database.
Also see \fBAccountingStorageLoc\fR and \fBJobCompLoc\fR.

.TP
\fBDefaultStoragePass\fR
The password used to gain access to the database to store the
accounting and job completion data.
Only used for database type storage plugins, ignored otherwise.
Also see \fBAccountingStoragePass\fR and \fBJobCompPass\fR.

.TP
\fBDefaultStoragePort\fR
The listening port of the accounting storage and/or job completion
database server.
Only used for database type storage plugins, ignored otherwise.
Also see \fBAccountingStoragePort\fR and \fBJobCompPort\fR.

.TP
\fBDefaultStorageType\fR
The accounting and job completion storage mechanism type.  Acceptable
values at present include "filetxt", "mysql" and "none".
The value "filetxt" indicates that records will be written to a file.
The value "mysql" indicates that accounting records will be written to a MySQL
or MariaDB database.
The default value is "none", which means that records are not maintained.
Also see \fBAccountingStorageType\fR and \fBJobCompType\fR.

.TP
\fBDefaultStorageUser\fR
The user account for accessing the accounting storage and/or job
completion database.
Only used for database type storage plugins, ignored otherwise.
Also see \fBAccountingStorageUser\fR and \fBJobCompUser\fR.

.TP
\fBDisableRootJobs\fR
If set to "YES" then user root will be prevented from running any jobs.
The default value is "NO", meaning user root will be able to execute jobs.
\fBDisableRootJobs\fR may also be set by partition.

.TP
\fBEioTimeout\fR
The number of seconds srun waits for slurmstepd to close the TCP/IP
connection used to relay data between the user application and srun
when the user application terminates. The default value is 60 seconds.
May not exceed 65533.

.TP
\fBEnforcePartLimits\fR
If set to "YES" then jobs which exceed a partition's size and/or time limits
will be rejected at submission time. If set to "NO" then the job will be
accepted and remain queued until the partition limits are altered.
The default value is "NO".
NOTE: If set, then a job's QOS can not be used to exceed partition limits.

.TP
\fBEpilog\fR
Fully qualified pathname of a script to execute as user root on every
node when a user's job completes (e.g. "/usr/local/slurm/epilog"). A
glob pattern (See \fBglob\fR (7)) may also be used to run more than
one epilog script (e.g. "/etc/slurm/epilog.d/*"). The Epilog script
or scripts may be used to purge files, disable user login, etc.
By default there is no epilog.
See \fBProlog and Epilog Scripts\fR for more information.

.TP
\fBEpilogMsgTime\fR
The number of microseconds that the slurmctld daemon requires to process
an epilog completion message from the slurmd daemons. This parameter can
be used to prevent a burst of epilog completion messages from being sent
at the same time which should help prevent lost messages and improve
throughput for large jobs.
The default value is 2000 microseconds.
For a 1000 node job, this spreads the epilog completion messages out over
two seconds.

.TP
\fBEpilogSlurmctld\fR
Fully qualified pathname of a program for the slurmctld to execute
upon termination of a job allocation (e.g.
"/usr/local/slurm/epilog_controller").
The program executes as SlurmUser, which gives it permission to drain
nodes and requeue the job if a failure occurs (See scontrol(1)).
Exactly what the program does and how it accomplishes this is completely at
the discretion of the system administrator.
Information about the job being initiated, it's allocated nodes, etc. are
passed to the program using environment variables.
See \fBProlog and Epilog Scripts\fR for more information.

.TP
\fBExtSensorsFreq\fR
The external sensors plugin sampling interval.
If \fBExtSensorsType=ext_sensors/none\fR, this parameter is ignored.
For all other values of \fBExtSensorsType\fR, this parameter is the number
of seconds between external sensors samples for hardware components (nodes,
switches, etc.) The default value is zero. This value disables external
sensors sampling. Note: This parameter does not affect external sensors
data collection for jobs/steps.

.TP
\fBExtSensorsType\fR
Identifies the plugin to be used for external sensors data collection.
Slurmctld calls this plugin to collect external sensors data for jobs/steps
and hardware components. In case of node sharing between jobs the reported
values per job/step (through sstat or sacct) may not be accurate.  See also
"man ext_sensors.conf".

Configurable values at present are:
.RS
.TP 20
\fBext_sensors/none\fR
No external sensors data is collected.
.TP
\fBext_sensors/rrd\fR
External sensors data is collected from the RRD database.
.RE

.TP
\fBFairShareDampeningFactor\fR
Dampen the effect of exceeding a user or group's fair share of allocated
resources. Higher values will provides greater ability to differentiate
between exceeding the fair share at high levels (e.g. a value of 1 results
in almost no difference between overconsumption by a factor of 10 and 100,
while a value of 5 will result in a significant difference in priority).
The default value is 1.

.TP
\fBFastSchedule\fR
Controls how a node's configuration specifications in slurm.conf are used.
If the number of node configuration entries in the configuration file
is significantly lower than the number of nodes, setting FastSchedule to
1 will permit much faster scheduling decisions to be made.
(The scheduler can just check the values in a few configuration records
instead of possibly thousands of node records.)
Note that on systems with hyper\-threading, the processor count
reported by the node will be twice the actual processor count.
Consider which value you want to be used for scheduling purposes.
.RS
.TP 5
\fB0\fR
Base scheduling decisions upon the actual configuration of each individual
node except that the node's processor count in SLURM's configuration must
match the actual hardware configuration if \fBSchedulerType=sched/gang\fR
or \fBSelectType=select/cons_res\fR are configured (both of those plugins
maintain resource allocation information using bitmaps for the cores in the
system and must remain static, while the node's memory and disk space can
be established later).
.TP
\fB1\fR (default)
Consider the configuration of each node to be that specified in the
slurm.conf configuration file and any node with less than the
configured resources will be set to DRAIN.
.TP
\fB2\fR
Consider the configuration of each node to be that specified in the
slurm.conf configuration file and any node with less than the
configured resources will \fBnot\fR be set DRAIN.
This option is generally only useful for testing purposes.
.RE

.TP
\fBFirstJobId\fR
The job id to be used for the first submitted to SLURM without a
specific requested value. Job id values generated will incremented by 1
for each subsequent job. This may be used to provide a meta\-scheduler
with a job id space which is disjoint from the interactive jobs.
The default value is 1.
Also see \fBMaxJobId\fR

.TP
\fBGetEnvTimeout\fR
Used for Moab scheduled jobs only. Controls how long job should wait
in seconds for loading the user's environment before attempting to
load it from a cache file. Applies when the srun or sbatch

Version-less_number_of_pull_get/PE-7-July/cons_res/exp1/riplay/SLURM.rb
#!/usr/bin/ruby
 
require 'pp'
require "./lib_swf.rb"
require "./JOB.rb"

def sbatch(job, type, output_dir, schedule_now, use_users, use_energy)
    job_id = job.job_id
    construct_script_slurm(job, type, output_dir, use_users, use_energy)
    command = "sbatch"
    if(!schedule_now)
       command = "#{command} --hold" 
    end
    command = "#{command} #{output_dir}/#{job_id}.slurm.job"
    puts "# bash -c '#{command}'"
    print `bash -c '#{command}'`
end

def slurmresume(jobs)
    running_jobs = `squeue -a | awk -F " "  '{print $1}' | grep -v 'JOBID'`
    dodo = 0
    running_jobs.each do |job_id|
        command = "scontrol release #{job_id}"   # TODO: here we release all the jobs, need to do it properly
        puts "# bash -c '#{command}'"
	print `bash -c '#{command}'`
        dodo = dodo + 1
        if(dodo == 20)
            dodo = 0
            sleep(0.01)
        end
    end
end


def construct_script_slurm(job, type, output_dir, use_users, use_energy)
    if(job.run_time_req == -1) 
       job.run_time_req = 3600 
    end    
    walltime = sec_to_hms(job.run_time_req)
    #nb_procs = job.procs_req # use procs_alloc instead of proc_req as this field is sometimes empty in models (lublin) or in slurm
    nb_procs = job.procs_alloc
    duration = job.run_time
    job_id = job.job_id

    puts "# create sbatch file #{job_id} walltime: #{walltime} nbprocs: #{nb_procs}"
    
    File.open("#{output_dir}/#{job_id}.slurm.job", 'w') do |f|
        f.puts "#\!/bin/sh"
        f.puts "#SBATCH -J riplay_#{job_id}_#{type}"
	f.puts "#SBATCH -n #{nb_procs}"
        f.puts "#SBATCH -t #{walltime}"
        f.puts "#SBATCH -o #{output_dir}/riplay_slurm_#{type}_#{job_id}.out"

	if use_users
		user_id = job.user_id
		if(user_id == 3)
			f.puts "#SBATCH --uid=50108"
			freq = 2394000
		elsif(user_id == 1)
			f.puts "#SBATCH --uid=50241"
			freq = 2527000
		elsif(user_id == 2)
			f.puts "#SBATCH --uid=50243"
			freq = 1729000
		end
	end
	if use_energy
		energy = job.energy
		f.puts "#SBATCH --comment=\"energy:#{energy}\"" 
	end

        f.puts ""
        #f.puts "srun --acctg-freq=energy=1 --exclusive --cpu-freq=#{freq} #{get_job_code(job, type, output_dir, 'SLURM', use_users)}"
	f.puts "srun #{get_job_code(job, type, output_dir, 'SLURM', use_users)}"
	f.puts "srun /bin/hostname"
        f.puts "sleep 1"
        f.puts "exit 0"
        f.puts ""  

    end
    `chmod 777 #{output_dir}/#{job_id}.slurm.job`
end


def slurmmake_energy_resv(time, duration, watts)
	d = (duration.to_i/60).to_i
	command = "scontrol create res FLAG=LICENSE_ONLY starttime=now+#{time} duration=#{d} Watts=#{watts} Users=root"
	puts "# bash -c '#{command}'"
	print `bash -c '#{command}'`
end
